### ISSUE TITLE
Repo: Consider a lint rule against multiple errors across multiple lines in tests

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#6664]
### Suggestion

I've been asking folks to keep each individual test case in `eslint-plugin`'s `tests/rules/**/*.test.ts` test files to having approximately one error each. Having multiple lines of lint rule failures in a single test case is oftentimes a sign that the test is testing multiple things.

Example: https://github.com/typescript-eslint/typescript-eslint/blob/0d0639fb25e5ac8d7b5e4ceac4273e17e8ee0249/packages/eslint-plugin/tests/rules/consistent-type-assertions.test.ts#L146

Proposal: how about we add a lint rule against `invalid` test cases having `errors` with multiple entries? There can be an exemption for entries that have the same `line` explicitly indicated, for cases of rules where multiple failures on a single expression should be tested.

**Comment by bradzacher** [comment]
`batchedSingleLineTests` was a util I created on purpose to help reduce some boilerplate.

in the past I've seen people do patterns like
```ts
const invalidCases = [
  'a',
  'b',
  'c',
  'd',
  // ...
] as const;

ruleTester.run('rule', rule, {
  invalid: [
    ...invalidCases.map(code => ({
      code,
      errors: [{ messageId: 'foo' }],
    }),
  ],
});
```

Or like:
```ts
ruleTester.run('rule', rule, {
  invalid: [
    {
      code: `
a
b
c
d
// ...
      `,
      errors: [
        { messageId: 'foo' },
        { messageId: 'foo' },
        { messageId: 'foo' },
        { messageId: 'foo' },
        // ...
      ],
    },
  ],
});
```

Both cases I never liked because the former can be hard to type correctly, and the latter is a bad test because if one fails they all fail and it's hard to figure out which one was the cause of the failure.

`batchedSingleLineTests` was supposed to be a middle ground between the two like
```ts
ruleTester.run('rule', rule, {
  invalid: [
    ...batchedSingleLineTests({
      code: `
a
b
c
d
// ...
      `,
      errors: [
        { messageId: 'foo', line: 2 },
        { messageId: 'foo', line: 3 },
        { messageId: 'foo', line: 4 },
        { messageId: 'foo', line: 5 },
        // ...
      ],
    }),
  ],
});
```

Under-the hood it translates this test case to this
```ts
ruleTester.run('rule', rule, {
  invalid: [
    {
      code: `a`,
      errors: [{ messageId: 'foo' }],
    },
    {
      code: `b`,
      errors: [{ messageId: 'foo' }],
    },
    {
      code: `c`,
      errors: [{ messageId: 'foo' }],
    },
    {
      code: `d`,
      errors: [{ messageId: 'foo' }],
    },
    // ...
  ],
});
```

If one case fails you instead get a failure for just that one line - and you also have the benefit of not duplicating as much boilerplate for the tests.

**Comment by bradzacher** [comment]
outside `batchedSingleLineTests` - generally yeah multiple errors per test is bad. There are some cases where you want to do this testing though to ensure that you are correctly reporting multiple cases.

For example sometimes you might add caching/short-circuiting to a rule to bail out of repeated processing, and want to ensure specific cases aren't short-circuited and get a report.
